File: day_16/p2.py
from collections import deque
from enum import Enum
from typing import Dict, Generator, List, Tuple
import functools

class DIR(Enum):
    NORTH = 0,
    SOUTH = 1,
    EAST = 2,
    WEST = 3

def parse(fname: str) -> List[str]:
    lines = []
    with open(fname, 'r') as f:
        lines = f.read().splitlines()

    return lines

def move_gen(cx: int, cy: int) -> Generator[Tuple[int, int, DIR], None, None]:
    yield -1,  0, DIR.NORTH
    yield  1,  0, DIR.SOUTH
    yield  0,  1, DIR.EAST
    yield  0, -1, DIR.WEST

def dijkstra(grid: List[str],
             start: Tuple[int, int, DIR],
             end: Tuple[int, int],
             score: int,
) -> int:
    x, y, hdg = start
    visited: Dict[Tuple[int, int, DIR], int] = {}
    results = []
    
    q = deque([ ((x, y, hdg, [(x,y)]), score) ])
    while q:
        (cx, cy, hdg, history), score = q.popleft()

        if (cx, cy) == end:
            results.append({score : history})
            continue

        state = (cx, cy, hdg)
        if state in visited and visited[state] < score:
            continue
        visited[state] = score
        
        for dx, dy, direction in move_gen(cx, cy):
            nx = cx + dx
            ny = cy + dy
            if nx in range(len(grid)) and ny in range(len(grid[0])) and grid[nx][ny] != '#':
                if (direction == hdg):
                    nhistory = history + [(nx, ny)]
                    q.append(( (nx, ny, hdg, nhistory), score + 1 ))
                else:
                    nhistory = history + []
                    q.append(( (cx, cy, direction, nhistory), score + 1000))
    
    unique_tiles = set()
    for result in results:
        for score, history in result.items():
            if score == min(key for result in results for key in result.keys()):
                unique_tiles.update(history)

    return len(unique_tiles)
    
def compute(grid: List[str]) -> int:
    n, m = len(grid), len(grid[0])
    start_x = start_y = 0

    for i in range(n):
        for j in range(m):
            if grid[i][j] == 'S':
                start = (i, j, DIR.EAST)
            elif grid[i][j] == 'E':
                end = (i, j)

    score = dijkstra(grid, start, end, 0)
    return score

def main() -> None:
    grid = parse("input.txt")
    print(compute(grid))


if __name__ == "__main__":
    main()

File: day_16/p2.cpp
#include "pypp.hpp" // https://github.com/A-Bergonzoli/pypp/tree/main

#include <iostream>

constexpr auto TARGET_COST { 105508 };

struct Move {
    int dx;
    int dy;
    grid::Direction hdg;
};

strings Parse(const std::string& fname)
{
    return pypp::splitFileLines(fname);
}

std::array<std::pair<Move, int>, 4U> Generator(grid::Direction hdg)
{
    std::array<std::pair<Move, int>, 4U> moves;
    const Move standard_moves[] = {
        { -1, 0, grid::Direction::UP },
        { 1, 0, grid::Direction::DOWN },
        { 0, 1, grid::Direction::RIGHT },
        { 0, -1, grid::Direction::LEFT }
    };

    for (std::size_t i { 0U }; i < moves.size(); ++i) {
        auto candidate = standard_moves[i];
        auto cost = 1;
        if (hdg != standard_moves[i].hdg) {
            candidate.dx = 0;
            candidate.dy = 0;
            cost = 1000;
        }
        moves[i] = std::make_pair(candidate, cost);
    }

    return moves;
}

int Compute(const strings& grid)
{
    int n = grid.size(), m = grid[0].length();
    grid::LocationWithHeading start(0, 0, grid::Direction::RIGHT);
    grid::Location end(0, 0);

    for (std::size_t i { 0U }; i < n; ++i) {
        for (std::size_t j { 0U }; j < m; ++j) {
            if (grid[i][j] == 'S')
                start = std::make_tuple(i, j, grid::Direction::RIGHT);
            else if (grid[i][j] == 'E')
                end = std::make_pair(i, j);
        }
    }

    auto [min_score, visited, best_paths] = algorithms::dijkstra_extended(grid, start, end, Generator, TARGET_COST);

    std::unordered_set<grid::Location, collections::TupleHash> unique_tiles {};
    for (auto path : best_paths) {
        for (auto elem : path)
            unique_tiles.insert({ std::get<0>(elem), std::get<1>(elem) });
    }

    return unique_tiles.size();
}

int main()
{
    const auto grid = Parse("input.txt");
    std::cout << Compute(grid) << std::endl;

    return 0;
}
